### DISCUSSION TITLE
differentiating identical event-services from a single client

### DISCUSSION THREAD

**Discussion by mogar** [COVESA/vsomeip#628] [Q&A] [answered]
I have two of the same some/ip devices. Each device acts as a service that can send various event notifications. They're identical except that they each have a different (static) IP address. So far I have tested event receiving with a single device, and it works well.

I'd like to be able to subscribe to events from both of the devices using one client. In the receiving client, what is the best way to differentiate which of the devices sent a given event?

**Comment by BrunoLSilvaCTW** [comment]
Hi @mogar!

The way to differentiate between events of the same service, but sent from different devices, is by using `message->get_instance()`. This `instance` is unique for every copy of a service in the vehicle network, and it's how you map a service to a specific device.

To subscribe to the events of both devices, you need to specify the `instance_id` of each service that you are interested in:

```cpp
app_->subscribe(SERVICE_ID, INSTANCE_ID_A, EVENTGROUP);
app_->subscribe(SERVICE_ID, INSTANCE_ID_B, EVENTGROUP);
```

To my knowledge, there is no away to get the ip address used by a given service instance at runtime. Ideally, this is something you should know beforehand.

Does this answer your question?

**Reply by DiogoPedrozza** [reply, marked as answer]
For further documentation please search the user guide (https://github.com/COVESA/vsomeip/blob/master/documentation/vsomeipUserGuide.md) for "instance" as there are several descriptions that could assist your implementation, also there is several examples of configuration files at: https://github.com/COVESA/vsomeip/tree/master/config

**Reply by nkuang123** [reply]
@BrunoLSilvaCTW Suppose we'd like to differentiate identical events with the same service ID and instance ID coming from multiple devices—how would the application disambiguate without the IP address?

**Reply by BrunoLSilvaCTW** [reply]
I don't know if you can. Each instance id must be unique per device. That's its purpose, to differentiate identical events with the same service id coming from multiple devices.

**Reply by siggie0815** [reply]
Hello,
actually, I do have a very similar problem: I do have some prototype sensor devices using SOME/IP with Service-Discovery for communication. Unfortunately, it is not possible to change the Instance ID on the sensors in their current firmware revision. Hence, the only way I can distinguish the data streams would be the IP Address.

I know that this is somehow contradicting the Service Discovery philosophy but I need to work with what I have. So, would it be somehow possible to use the IP addresses? Maybe by manually defining the service endpoints or something?

I already thought on how I could make it possible: My first Idea was to extend the `vsomeip_v3::instance_t` to be a tuple of IP and ID, so I could define the IP when requesting services and I could get the IP when receiving data. However, I guess this would not be a "small workaround" 😆

So, I would be very grateful for any Idea on how I could do it without this "workaround"?
